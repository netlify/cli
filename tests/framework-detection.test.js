const test = require('ava')
const execa = require('execa')

const cliPath = require('./utils/cli-path')
const { getExecaOptions, withDevServer } = require('./utils/dev-server')
const got = require('./utils/got')
const { CONFIRM, DOWN, handleQuestions } = require('./utils/handle-questions')
const { withSiteBuilder } = require('./utils/site-builder')
const { normalize } = require('./utils/snapshots')

const content = 'Hello World!'

test('should default to process.cwd() and static server', async (t) => {
  await withSiteBuilder('site-with-index-file', async (builder) => {
    await builder
      .withContentFile({
        path: 'index.html',
        content,
      })
      .buildAsync()

    await withDevServer({ cwd: builder.directory }, async ({ output, url }) => {
      const response = await got(url).text()
      t.is(response, content)

      t.snapshot(normalize(output))
    })
  })
})

test('should use static server when --dir flag is passed', async (t) => {
  await withSiteBuilder('site-with-index-file', async (builder) => {
    await builder
      .withContentFile({
        path: 'public/index.html',
        content,
      })
      .buildAsync()

    await withDevServer({ cwd: builder.directory, args: ['--dir', 'public'] }, async ({ output, url }) => {
      const response = await got(url).text()
      t.is(response, content)

      t.snapshot(normalize(output))
    })
  })
})

test('should use static server when framework is set to #static', async (t) => {
  await withSiteBuilder('site-with-index-file', async (builder) => {
    await builder
      .withContentFile({
        path: 'index.html',
        content,
      })
      .withNetlifyToml({ config: { dev: { framework: '#static' } } })
      .buildAsync()

    await withDevServer({ cwd: builder.directory }, async ({ output, url }) => {
      const response = await got(url).text()
      t.is(response, content)

      t.snapshot(normalize(output))
    })
  })
})

test('should warn if using static server and `command` is configured', async (t) => {
  await withSiteBuilder('site-with-index-file', async (builder) => {
    await builder
      .withContentFile({
        path: 'public/index.html',
        content,
      })
      .buildAsync()

    await withDevServer(
      { cwd: builder.directory, args: ['--dir', 'public', '--command', 'npm run start'] },
      async ({ output, url }) => {
        const response = await got(url).text()
        t.is(response, content)

        t.snapshot(normalize(output))
      },
    )
  })
})

test('should warn if using static server and `targetPort` is configured', async (t) => {
  await withSiteBuilder('site-with-index-file', async (builder) => {
    await builder
      .withContentFile({
        path: 'public/index.html',
        content,
      })
      .buildAsync()

    await withDevServer(
      { cwd: builder.directory, args: ['--dir', 'public', '--targetPort', '3000'] },
      async ({ output, url }) => {
        const response = await got(url).text()
        t.is(response, content)

        t.snapshot(normalize(output))
      },
    )
  })
})

test('should run `command` when both `command` and `targetPort` are configured', async (t) => {
  await withSiteBuilder('empty-site', async (builder) => {
    await builder.withNetlifyToml({ config: { build: { publish: 'public' } } }).buildAsync()

    // a failure is expected since we use `echo hello` instead of starting a server
    const error = await t.throwsAsync(() =>
      withDevServer(
        { cwd: builder.directory, args: ['--command', 'echo hello', '--targetPort', '3000'] },
        () => {},
        true,
      ),
    )
    t.snapshot(normalize(error.stdout))
  })
})

test('should force a specific framework when configured', async (t) => {
  await withSiteBuilder('site-with-mocked-cra', async (builder) => {
    await builder.withNetlifyToml({ config: { dev: { framework: 'create-react-app' } } }).buildAsync()

    // a failure is expected since this is not a true create-react-app project
    const error = await t.throwsAsync(() => withDevServer({ cwd: builder.directory }, () => {}, true))
    t.snapshot(normalize(error.stdout))
  })
})

test('should throw when forcing a non supported framework', async (t) => {
  await withSiteBuilder('site-with-unknown-framework', async (builder) => {
    await builder.withNetlifyToml({ config: { dev: { framework: 'to-infinity-and-beyond-js' } } }).buildAsync()

    const error = await t.throwsAsync(() => withDevServer({ cwd: builder.directory }, () => {}, true))
    t.snapshot(normalize(error.stdout))
  })
})

test('should detect a known framework', async (t) => {
  await withSiteBuilder('site-with-cra', async (builder) => {
    await builder
      .withPackageJson({
        packageJson: { dependencies: { 'react-scripts': '1.0.0' }, scripts: { start: 'react-scripts start' } },
      })
      .buildAsync()

    // a failure is expected since this is not a true create-react-app project
    const error = await t.throwsAsync(() => withDevServer({ cwd: builder.directory }, () => {}, true))
    t.snapshot(normalize(error.stdout))
  })
})

test('should throw if framework=#custom but command is missing', async (t) => {
  await withSiteBuilder('site-with-framework-and-no-command', async (builder) => {
    await builder.withNetlifyToml({ config: { dev: { framework: '#custom' } } }).buildAsync()

    const error = await t.throwsAsync(() =>
      withDevServer({ cwd: builder.directory, args: ['--targetPort', '3000'] }, () => {}, true),
    )
    t.snapshot(normalize(error.stdout))
  })
})

test('should throw if framework=#custom but targetPort is missing', async (t) => {
  await withSiteBuilder('site-with-framework-and-no-command', async (builder) => {
    await builder.withNetlifyToml({ config: { dev: { framework: '#custom' } } }).buildAsync()

    const error = await t.throwsAsync(() =>
      withDevServer({ cwd: builder.directory, args: ['--command', 'echo hello'] }, () => {}, true),
    )
    t.snapshot(normalize(error.stdout))
  })
})

test('should start custom command if framework=#custom, command and targetPort are configured', async (t) => {
  await withSiteBuilder('site-with-custom-framework', async (builder) => {
    await builder.withNetlifyToml({ config: { dev: { framework: '#custom', publish: 'public' } } }).buildAsync()

    const error = await t.throwsAsync(() =>
      withDevServer(
        { cwd: builder.directory, args: ['--command', 'echo hello', '--targetPort', '3000'] },
        () => {},
        true,
      ),
    )
    t.snapshot(normalize(error.stdout))
  })
})

test(`should print specific error when command doesn't exist`, async (t) => {
  await withSiteBuilder('site-with-custom-framework', async (builder) => {
    await builder.buildAsync()

    const error = await t.throwsAsync(() =>
      withDevServer(
        {
          cwd: builder.directory,
          args: [
            '--command',
            'oops-i-did-it-again forgot-to-use-a-valid-command',
            '--targetPort',
            '3000',
            '--framework',
            '#custom',
          ],
        },
        () => {},
        true,
      ),
    )
    t.snapshot(normalize(error.stdout))
  })
})

test('should prompt when multiple frameworks are detected', async (t) => {
  await withSiteBuilder('site-with-multiple-frameworks', async (builder) => {
    await builder
      .withPackageJson({
        packageJson: {
          dependencies: { 'react-scripts': '1.0.0', gatsby: '^3.0.0' },
          scripts: { start: 'react-scripts start', develop: 'gatsby develop' },
        },
      })
      .withContentFile({ path: 'gatsby-config.js', content: '' })
      .buildAsync()

    // a failure is expected since this is not a true framework project
    const error = await t.throwsAsync(async () => {
      const childProcess = execa(cliPath, ['dev', '--offline'], getExecaOptions({ cwd: builder.directory }))

      handleQuestions(childProcess, [
        {
          question: 'Multiple possible start commands found',
          answer: `${DOWN}${CONFIRM}`,
        },
      ])

      await childProcess
    })
    t.snapshot(normalize(error.stdout))
  })
})

test('should not run framework detection if command and targetPort are configured', async (t) => {
  await withSiteBuilder('site-with-hugo-config', async (builder) => {
    await builder.withContentFile({ path: 'config.toml', content: '' }).buildAsync()

    // a failure is expected since the command exits early
    const error = await t.throwsAsync(() =>
      withDevServer(
        { cwd: builder.directory, args: ['--command', 'echo hello', '--targetPort', '3000'] },
        () => {},
        true,
      ),
    )
    t.snapshot(normalize(error.stdout))
  })
})

test('should filter frameworks with no dev command', async (t) => {
  await withSiteBuilder('site-with-gulp', async (builder) => {
    await builder
      .withContentFile({
        path: 'index.html',
        content,
      })
      .withPackageJson({
        packageJson: { dependencies: { gulp: '1.0.0' } },
      })
      .buildAsync()

    await withDevServer({ cwd: builder.directory }, async ({ output, url }) => {
      const response = await got(url).text()
      t.is(response, content)

      t.snapshot(normalize(output))
    })
  })
})

test('should pass framework-info env to framework sub process', async (t) => {
  await withSiteBuilder('site-with-gatsby', async (builder) => {
    await builder
      .withPackageJson({
        packageJson: {
          dependencies: { gatsby: '^3.0.0' },
          scripts: { develop: 'node -p process.env.GATSBY_LOGGER' },
        },
      })
      .withContentFile({ path: 'gatsby-config.js', content: '' })
      .buildAsync()

    // a failure is expected since this is not a true Gatsby project
    const error = await t.throwsAsync(() => withDevServer({ cwd: builder.directory }, () => {}, true))
    t.snapshot(normalize(error.stdout))
  })
})
