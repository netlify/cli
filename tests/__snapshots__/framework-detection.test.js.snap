// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should default to process.cwd() and static server 1`] = `
"◈ Netlify Dev ◈
◈ No app server detected. Using simple static server
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://cli.netlify.com/netlify-dev#project-detection
◈ Running static server from \\"site-with-index-file\\"

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘

"
`;

exports[`should detect a known framework 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888
sh: react-scripts: command not found
◈ Netlify Dev ◈
◈ Starting Netlify Dev with Create React App

> start
> react-scripts start

◈ Command failed with exit code 127: npm run start. Shutting down Netlify Dev server"
`;

exports[`should filter frameworks with no dev command 1`] = `
"◈ Netlify Dev ◈
◈ No app server detected. Using simple static server
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://cli.netlify.com/netlify-dev#project-detection
◈ Running static server from \\"site-with-gulp\\"

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘

"
`;

exports[`should force a specific framework when configured 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888
◈ Netlify Dev ◈
◈ Starting Netlify Dev with Create React App
◈ Failed running command: react-scripts start. Please verify 'react-scripts' exists"
`;

exports[`should log the command if using static server and \`command\` is configured 1`] = `
"◈ Netlify Dev ◈
◈ Using simple static server because '--dir' flag was specified
◈ Running static server from \\"site-with-index-file/public\\"

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘

"
`;

exports[`should not run framework detection if command and targetPort are configured 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888 --command echo hello --targetPort 3000
◈ Netlify Dev ◈
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://cli.netlify.com/netlify-dev#project-detection
◈ Starting Netlify Dev with custom config
hello
◈ \\"echo hello\\" exited with code 0. Shutting down Netlify Dev server"
`;

exports[`should pass framework-info env to framework sub process 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888
◈ Netlify Dev ◈
◈ Starting Netlify Dev with Gatsby

> develop
> node -p process.env.GATSBY_LOGGER

yurnalist
◈ \\"npm run develop\\" exited with code 0. Shutting down Netlify Dev server"
`;

exports[`should print specific error when command doesn't exist 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888 --command oops-i-did-it-again forgot-to-use-a-valid-command --targetPort 3000 --framework #custom
◈ Netlify Dev ◈
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://cli.netlify.com/netlify-dev#project-detection
◈ Starting Netlify Dev with #custom
◈ Failed running command: oops-i-did-it-again forgot-to-use-a-valid-command. Please verify 'oops-i-did-it-again' exists"
`;

exports[`should prompt when multiple frameworks are detected 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline
sh: react-scripts: command not found
◈ Netlify Dev ◈
? Multiple possible start commands found (Use arrow keys or type to search)
> [Gatsby] 'npm run develop' 
  [Create React App] 'npm run start' ? Multiple possible start commands found 
  [Gatsby] 'npm run develop' 
> [Create React App] 'npm run start' ? Multiple possible start commands found Create React App-npm run start
Add 'framework = \\"create-react-app\\"' to the [dev] section of your netlify.toml to avoid this selection prompt next time
◈ Starting Netlify Dev with Create React App

> start
> react-scripts start

◈ Command failed with exit code 127: npm run start. Shutting down Netlify Dev server"
`;

exports[`should run \`command\` when both \`command\` and \`targetPort\` are configured 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888 --command echo hello --targetPort 3000
◈ Netlify Dev ◈
◈ Starting Netlify Dev with custom config
hello
◈ \\"echo hello\\" exited with code 0. Shutting down Netlify Dev server"
`;

exports[`should start custom command if framework=#custom, command and targetPort are configured 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888 --command echo hello --targetPort 3000
◈ Netlify Dev ◈
◈ Starting Netlify Dev with #custom
hello
◈ \\"echo hello\\" exited with code 0. Shutting down Netlify Dev server"
`;

exports[`should throw if framework=#custom but command is missing 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888 --targetPort 3000
◈ Netlify Dev ◈
◈ 'command' and 'targetPort' properties are required when 'framework' is set to '#custom'"
`;

exports[`should throw if framework=#custom but targetPort is missing 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888 --command echo hello
◈ Netlify Dev ◈
◈ 'command' and 'targetPort' properties are required when 'framework' is set to '#custom'"
`;

exports[`should throw when forcing a non supported framework 1`] = `
"Command failed with exit code *: /Users/lukasholzer/Sites/netlify/cli/bin/run dev --offline -p 88888 --staticServerPort 88888
◈ Netlify Dev ◈
◈ Invalid framework \\"to-infinity-and-beyond-js\\". It should be one of: *"
`;

exports[`should use static server when --dir flag is passed 1`] = `
"◈ Netlify Dev ◈
◈ Using simple static server because '--dir' flag was specified
◈ Running static server from \\"site-with-index-file/public\\"

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘

"
`;

exports[`should use static server when framework is set to #static 1`] = `
"◈ Netlify Dev ◈
◈ Using simple static server because '[dev.framework]' was set to '#static'
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://cli.netlify.com/netlify-dev#project-detection
◈ Running static server from \\"site-with-index-file\\"

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘

"
`;

exports[`should warn if using static server and \`targetPort\` is configured 1`] = `
"◈ Netlify Dev ◈
◈ Using simple static server because '--dir' flag was specified
◈ Ignoring 'targetPort' setting since using a simple static server.
◈ Use --staticServerPort or [dev.staticServerPort] to configure the static server port
◈ Running static server from \\"site-with-index-file/public\\"

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘

"
`;
